### PR TITLE
Issue #18945. Fix HandOff problem where Shards may not be killed.

### PR DESCRIPTION
**Pull request by msosnicki**

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @msosnicki,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.typesafe.com/contribute/cla](http://www.typesafe.com/contribute/cla)

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
Resolves https://github.com/akka/akka/issues/18945

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala [hunk 1]
```diff
@@ -722,29 +724,35 @@ class ShardRegion(
...
-          name))
-        shardsByRef = shardsByRef.updated(shard, id)
-        None
-      case Some(props) ⇒
-        None
-      case None ⇒
-        throw new IllegalStateException("Shard must not be allocated to a proxy only ShardRegion")
-    })
+  def getShard(id: ShardId): Option[ActorRef] = startingShards.contains(id) match {
+    case true ⇒ None
```
Comment (drewhk) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: why match on a Boolean? I think it is more idiomatic to use `if-else` here.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala [hunk 2]
```diff
@@ -632,7 +634,7 @@ class ShardRegion(
 
   def initializeShard(id: ShardId, shard: ActorRef): Unit = {
     log.debug("Shard was initialized {}", id)
-    shards = shards.updated(id, shard)
+    startingShards -= id
```
Comment (drewhk) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: what does the name `staringShards` mean?

Comment (msosnicki) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: This set holds all the shards actors that have already been started, but not yet recovered.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala (hunk 1 above)
Comment (msosnicki) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: You're right, maybe it's more common to do the regular if-else. I'll change that.

**Comment by msosnicki** [comment]
@drewhk Thanks for having a look into this PR, it's updated now.

**Comment by patriknw** [comment]
LGTM, thanks for contributing, @msosnicki

**Comment by patriknw** [comment]
Refs #18945

**Pull request merged**